File: tui/internal/api/stats_test.go
package api

import "testing"

// The /stats body serve/stats.py emits (fields asserted by test_serve_app.py).
const statsJSON = `{"model":"target-fake","draft":"draft-fake","busy":true,"k_current":3,` +
	`"alpha_ewma":0.5,"tok_s_recent":41.25,"windows_total":9,"accepted_total":4,` +
	`"proposed_total":27,"last_windows":[` +
	`{"k_proposed":3,"n_accepted":3,"draft_ms":1.5,"verify_ms":9.0,"window_ms":11.0},` +
	`{"k_proposed":3,"n_accepted":1,"draft_ms":1.5,"verify_ms":9.0,"window_ms":11.0}]}`

func TestParseStats(t *testing.T) {
	got, err := ParseStats(statsJSON)
	if err != nil {
		t.Fatalf("ParseStats: %v", err)
	}

	if got.Model != "target-fake" || got.Draft != "draft-fake" {
		t.Errorf("model/draft = %q/%q", got.Model, got.Draft)
	}
	if !got.Busy || got.KCurrent != 3 {
		t.Errorf("busy=%v k=%d", got.Busy, got.KCurrent)
	}
	if got.AlphaEWMA == nil || *got.AlphaEWMA != 0.5 {
		t.Errorf("alpha = %v", got.AlphaEWMA)
	}
	if got.TokSRecent == nil || *got.TokSRecent != 41.25 {
		t.Errorf("tok/s = %v", got.TokSRecent)
	}
	if got.WindowsTotal != 9 || got.AcceptedTotal != 4 || got.ProposedTotal != 27 {
		t.Errorf("totals = %d %d %d", got.WindowsTotal, got.AcceptedTotal, got.ProposedTotal)
	}
	if len(got.LastWindows) != 2 {
		t.Fatalf("last_windows len = %d", len(got.LastWindows))
	}
	w := got.LastWindows[1]
	if w.KProposed != 3 || w.NAccepted != 1 || w.DraftMs != 1.5 || w.VerifyMs != 9 || w.WindowMs != 11 {
		t.Errorf("window[1] = %+v", w)
	}
}

func TestParseStatsNullsAndNoDraft(t *testing.T) {
	got, err := ParseStats(`{"model":"m","draft":null,"busy":false,"k_current":0,` +
		`"alpha_ewma":null,"tok_s_recent":null,"windows_total":0,"accepted_total":0,` +
		`"proposed_total":0,"last_windows":[]}`)
	if err != nil {
		t.Fatalf("ParseStats: %v", err)
	}

	if got.Draft != "" || got.AlphaEWMA != nil || got.TokSRecent != nil {
		t.Errorf("nulls not preserved: %+v", got)
	}
	if got.LastWindows == nil {
		t.Errorf("last_windows should be an empty slice, not nil")
	}
}

func TestParseStatsRejectsGarbage(t *testing.T) {
	if _, err := ParseStats(`{"model":`); err == nil {
		t.Fatal("want error")
	}
}

File: dashboard/src/lib/stats.test.ts
// The payload shape is the one tests/test_serve_app.py asserts on /stats and
// the `event: stats` frames of /stats/stream.
import { describe, expect, test } from 'vitest';

import { M4_V_BY_K } from './speedup';

import { StatsParseError, costRatio, parseStats } from './stats';
import type { WindowSummary } from './types';

const WINDOW = { k_proposed: 3, n_accepted: 2, draft_ms: 1.5, verify_ms: 4.0, window_ms: 6.2 };

const PAYLOAD = {
  model: 'target-fake',
  draft: 'draft-fake',
  busy: false,
  k_current: 3,
  alpha_ewma: 0.0,
  tok_s_recent: 41.7,
  windows_total: 9,
  accepted_total: 0,
  proposed_total: 27,
  last_windows: [WINDOW],
  v_by_k: M4_V_BY_K,
};

describe('parseStats', () => {
  test('accepts the exact /stats payload as a JSON string', () => {
    const stats = parseStats(JSON.stringify(PAYLOAD));
    expect(stats).toEqual(PAYLOAD);
  });

  test('accepts an already-decoded object', () => {
    expect(parseStats(PAYLOAD)).toEqual(PAYLOAD);
  });

  test('keeps the nullable fields null: idle server before any window', () => {
    const idle = { ...PAYLOAD, draft: null, alpha_ewma: null, tok_s_recent: null, last_windows: [] };
    const stats = parseStats(idle);
    expect(stats.draft).toBeNull();
    expect(stats.alpha_ewma).toBeNull();
    expect(stats.tok_s_recent).toBeNull();
    expect(stats.last_windows).toEqual([]);
  });

  test.each([
    ['not json', '{"model":'],
    ['missing field', { ...PAYLOAD, k_current: undefined }],
    ['wrong type', { ...PAYLOAD, busy: 'yes' }],
    ['bad window', { ...PAYLOAD, last_windows: [{ k_proposed: 'three' }] }],
    ['array root', []],
    ['null root', null],
  ])('rejects %s with StatsParseError', (_name, bad) => {
    expect(() => parseStats(bad)).toThrow(StatsParseError);
  });
});

describe('costRatio', () => {
  const w = (k: number, draft: number, verify: number): WindowSummary => ({
    k_proposed: k,
    n_accepted: 0,
    draft_ms: draft,
    verify_ms: verify,
    window_ms: draft + verify,
  });

  test('is the median over windows of per-drafted-token cost over verify cost', () => {
    // per-token costs: (1.2/4)/2 = 0.15, (2/4)/2.5 = 0.2, (0.3/3)/1 = 0.1 -> median 0.15
    expect(costRatio([w(4, 1.2, 2), w(4, 2, 2.5), w(3, 0.3, 1)])).toBeCloseTo(0.15, 12);
  });

  test('averages the middle pair for an even count', () => {
    expect(costRatio([w(4, 1.2, 2), w(4, 2, 2.5)])).toBeCloseTo(0.175, 12);
  });

  test('ignores baseline windows (k = 0) and windows without a verify time', () => {
    expect(costRatio([w(0, 0, 3), w(4, 1.2, 2), w(4, 1, 0)])).toBeCloseTo(0.15, 12);
  });

  test('is null when nothing is measurable yet', () => {
    expect(costRatio([])).toBeNull();
    expect(costRatio([w(0, 0, 3)])).toBeNull();
  });
});
